File: atlasos-sdk/routes/agent_health.js
/**
 * AtlasOS Agent Health Routes for Express.
 *
 * Exposes GET /agent/status and GET /agent/health for agent discovery and health checks.
 *
 * Environment:
 *   AGENT_STATE_FILE - Path to JSON state file (default: ./state/agent_state.json)
 *   AGENT_ID         - Agent identifier (fallback when state file absent)
 *   AGENT_NAME       - Human-readable agent name
 *   AGENT_VERSION    - Agent version string
 *   AGENT_CAPABILITIES - Comma-separated capabilities (optional)
 */

const fs = require("fs");
const path = require("path");

const AGENT_STATE_FILE = process.env.AGENT_STATE_FILE || "state/agent_state.json";
const AGENT_ID = process.env.AGENT_ID || "";
const AGENT_NAME = process.env.AGENT_NAME || "AtlasOS Agent";
const AGENT_VERSION = process.env.AGENT_VERSION || "1.0.0";
const AGENT_CAPABILITIES = (process.env.AGENT_CAPABILITIES || "")
  .split(",")
  .map((c) => c.trim())
  .filter(Boolean);

function loadState() {
  const fp = path.isAbsolute(AGENT_STATE_FILE)
    ? AGENT_STATE_FILE
    : path.join(process.cwd(), AGENT_STATE_FILE);
  if (!fs.existsSync(fp)) return {};
  try {
    return JSON.parse(fs.readFileSync(fp, "utf8"));
  } catch {
    return {};
  }
}

function agentInfo() {
  const state = loadState();
  return {
    agent_id: state.agent_id || AGENT_ID || "unknown",
    name: state.name || AGENT_NAME,
    version: state.version || AGENT_VERSION,
    capabilities: state.capabilities || AGENT_CAPABILITIES,
    status: state.status || "operational",
    last_heartbeat: state.last_heartbeat,
    started_at: state.started_at,
  };
}

/**
 * GET /agent/status - Agent info for discovery/registry
 */
function getAgentStatus(req, res) {
  const info = agentInfo();
  info.timestamp = new Date().toISOString();
  res.json(info);
}

/**
 * GET /agent/health - Health check for load balancers
 */
function getAgentHealth(req, res) {
  const info = agentInfo();
  const status = info.status || "operational";
  const healthy = ["operational", "ready", "running"].includes(status);

  res.json({
    status: healthy ? "healthy" : "degraded",
    agent_id: info.agent_id,
    version: info.version,
    timestamp: new Date().toISOString(),
    details: {
      operational_status: status,
      capabilities_count: (info.capabilities || []).length,
    },
  });
}

/**
 * Mount agent health routes on an Express app or router.
 * @param {import('express').Router|import('express').Application} app - Express app or router
 */
function mountAgentHealthRoutes(app) {
  app.get("/agent/status", getAgentStatus);
  app.get("/agent/health", getAgentHealth);
}

module.exports = {
  getAgentStatus,
  getAgentHealth,
  mountAgentHealthRoutes,
};

File: services/deploy/templates/repo-agents/routes/agent_health.py
"""
AtlasOS Agent Health Routes for FastAPI.

Exposes GET /agent/status and GET /agent/health for agent discovery and health checks.

Environment:
  AGENT_STATE_FILE - Path to JSON state file (default: ./state/agent_state.json)
  AGENT_ID         - Agent identifier (fallback when state file absent)
  AGENT_NAME       - Human-readable agent name
  AGENT_VERSION    - Agent version string
  AGENT_CAPABILITIES - Comma-separated capabilities (optional)
"""
from __future__ import annotations

import json
import os
from datetime import datetime, timezone
from pathlib import Path

from fastapi import APIRouter

router = APIRouter(prefix="/agent", tags=["agent"])

AGENT_STATE_FILE = os.environ.get("AGENT_STATE_FILE", "state/agent_state.json")
AGENT_ID = os.environ.get("AGENT_ID", "")
AGENT_NAME = os.environ.get("AGENT_NAME", "AtlasOS Agent")
AGENT_VERSION = os.environ.get("AGENT_VERSION", "1.0.0")
AGENT_CAPABILITIES = [
    c.strip() for c in (os.environ.get("AGENT_CAPABILITIES", "") or "").split(",") if c.strip()
]


def _load_state() -> dict:
    """Load agent state from file or return empty dict."""
    path = Path(AGENT_STATE_FILE)
    if not path.exists():
        return {}
    try:
        with path.open() as f:
            return json.load(f)
    except (json.JSONDecodeError, OSError):
        return {}


def _agent_info() -> dict:
    """Build agent info from state file or env vars."""
    state = _load_state()
    return {
        "agent_id": state.get("agent_id") or AGENT_ID or "unknown",
        "name": state.get("name") or AGENT_NAME,
        "version": state.get("version") or AGENT_VERSION,
        "capabilities": state.get("capabilities") or AGENT_CAPABILITIES,
        "status": state.get("status", "operational"),
        "last_heartbeat": state.get("last_heartbeat"),
        "started_at": state.get("started_at"),
    }


@router.get("/status")
async def get_agent_status() -> dict:
    """
    Return agent info from state file or env vars.
    Used for agent discovery and registry.
    """
    info = _agent_info()
    info["timestamp"] = datetime.now(timezone.utc).isoformat()
    return info


@router.get("/health")
async def get_agent_health() -> dict:
    """
    Return agent health status for load balancers and health checks.
    """
    info = _agent_info()
    status = info.get("status", "operational")
    healthy = status in ("operational", "ready", "running")

    return {
        "status": "healthy" if healthy else "degraded",
        "agent_id": info["agent_id"],
        "version": info["version"],
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "details": {
            "operational_status": status,
            "capabilities_count": len(info.get("capabilities", [])),
        },
    }
